### DISCUSSION TITLE
Ideas for Resolving Terminal Code Rot

### DISCUSSION THREAD

**Discussion by manchicken** [net-amqp-rabbitmq/net-amqp-rabbitmq#241] [Polls]
Hey all, I've been (poorly) maintaining this module for several years now, and I have to admit that we've run into some pretty serious code rot at this point. The dependencies (mostly OpenSSL) have moved faster than I've been able to, and at this point we're running on some deprecated stuff.

I have a few ideas on how we can resolve this, but there's a larger set of context I'd like to set out first.

## The big problems

1. The [BigInt module we use](https://github.com/salva/p5-Math-Int64) hasn't had an update in about eight years
2. The OpenSSL v1.1.1 that we depend on was EOL'ed in September of 2023, and we're stuck here because `librabbitmq-c` doesn't support any more recent versions.
3. There is so much copy-pasta from prior maintainers that we rely on files in `librabbitmq-c` which no longer exist
4. My time is remarkably limited
5. It's just me maintaining this thing

At this point, I kinda regard this module as terminally rotten, and in need of massive restoration. I've got a few paths ahead of me, and I'd like y'all's thoughts about this.

## The solutions I can think of

Here are the solutions I can think of, as well as their benefits and risks. Keep in mind, this will _not_ be a fast process. I've got a lot of obligations that take time, and I think that there's a high degree of risk that I might not finish without some support or at least cheerleading.

### 1. A Rust rewrite

This approach would mean figuring out how to use bindings in Rust rather than C, and using a well-maintained Rust library for the business end.

**Benefits**

- The RabbitMQ crate in Rust is pretty full-featured, so if I can get the bindings working the majority of the work will be type-mapping and binding to functions, but very little protocol work.
- Rust's ownership model will lead to more thread-safe and secure code.
- Confirm-select is already implemented in the Rust module, so if we could get these bindings to work we could add more functionality in that area.

**Risks**

- I don't know if binding to Rust is going to work without the cost of FFI.
- Rust types are very different from C types and Perl types, so there may be a lot more work there than anticipated.
- Rust's build environment is very different from a C build environment. It's simpler, but it's nowhere near as ubiquitous. We may lose compatibility with some OS's and chipsets as a result.

### 2. A C rewrite

A slash-and-burn rewrite of the C code we have now. In this approach I would necessarily remove all of the copy-pasta, and move to an approach that is a more traditional binding rather than doing so much protocol work in the XS code.

**Benefits**

- We already have a lot of well-tested type conversion code here, so we won't have to rewrite type conversion code.
- We can update dependencies and move closer to the supported logic of the `librabbitmq-c` library we use.

**Risks**

- The reason the copy-pasta is there is because it avoids type conversions to save performance. This means that there would likely be a performance degradation in this approach of an unknown quantity.
- New C code means new potential for security holes. Many security risks are born of buffer overflows. While I'm pretty handy with C, I am _not_ perfect, and security risk may arise (though the state of code rot already puts us in this position)
- `librabbitmq-c` does _not_ support OpenSSL beyond v1.1.1, so the risk of using a deprecated OpenSSL version would remain

### 3. Pure Perl rewrite

In Node-land, the `amqplib` module is pure-JS using code generated from the AMQP 0.9.1 spec. We could try this approach for the Perl module, too.

**Benefits**

- A pure-Perl approach would vastly improve compatibility
- We could jettison all of the dependencies and shoot for as few of them as possible (we wouldn't be able to remove all of them, though)
- This code would be inherently more testable
- A lot more thread safety and memory safety

**Risks**

- The performance of this version would likely be noticeably slower
- A number of dependencies in Perl land will likely also have some of the problems we're seeing already
- There will likely be breaking changes here, in order to support the new usage paradigm.
- A lot more work to update tests, I suspect

### 4. End-of-life this module

It is entirely possible that this module no longer adds value. If this is the case, I should stop investing time. If this is the path we take, then I will archive the repository and mark it as deprecated on CPAN.

I'm not going to go into benefits and risks, I feel like that's self-explanatory here.

## How you can help

- Contribute to code and testing as we move; especially if you have any code which exercises odd use-cases
- If you've got time, and you'd like to take over this project, please contact me

## The poll

I'd like y'all to weigh in on which direction I take. I will end this poll at on January 31st, 2024. If you have additional options, please add a comment.

I think that at the moment, my favorite approach is the Rust one, but that also seems like the most risky approach. If I find an obvious winner in this approach, I may scuttle the poll, but I won't do that unless it's an obvious improvement.

## Further discussion

As I do the work on this, I will continue using GitHub Discussions to keep y'all up-to-date. Please stick with me, though. It's no fun maintaining a module like this without any community involvement.

## Thanks

Thanks all for reading this, and I really hope to have some participation. I hope you have a great end to your 2023.

**Comment by manchicken** [comment]
Constructive comments are welcome.

**Comment by leejo** [comment]
There's been a pure perl version available for a while: https://metacpan.org/pod/Net::AMQP::RabbitMQ::PP # we took over maintenance of it a while back and uploaded it to CPAN as it had made it into our stack but wasn't there. I'm not sure the level of compatibility with this version, nor the performance compared to this version. It might be worth a look (patches welcome, although caveat emptor that we are not the original authors). See https://github.com/Humanstate/net-amqp-rabbitmq for the repo.

As to the C lib stuff - my colleague asks "where the EOL openssl requirement is, as the C library doesn’t seem to say that - is the C library ignoring a dirty secret?".

**Reply by manchicken** [reply]
Do you see the pure Perl module as being feature complete? If so, I'm happy to yield the namespace. 

I wonder how the performance differences are.

**Reply by leejo** [reply]
I'm not sure how feature complete it is - as I say we took over maintenance. We also only use a small subset of rabbitmq features. Performance differences I'm not sure about also.

**Comment by manchicken** [comment]
Hey all; here's an update!

A new release is likely coming soon. I've managed to upgrade the `rabbitmq-c` dependencies and get tests passing again. 

### What has kept you waiting?

Two things, really:

1. As discussed above, this module is not easy to work with. It's got some excessive complexity which makes it really risky to make changes on, and I hadn't been able to test it with OpenSSL 3 successfully
2. I am stupidly busy, and it wasn't until this week that I was able to set aside enough time with our old pal `gdb` to troubleshoot

### So where to we go from here?

In a couple of directions, but with goal nº1 being to reduce the time-suck of the toolchain.

1. I'm going to first focus on automating all of the tasks with this module
  a. Using GitHub Actions for CI
  b. Using locally-running scripts for publishing versions to PAUSE
2. I want to start a campaign of simplification in this codebase
3. I would like to attract collaborators. I don't want someone to hand off to, I want someone who can offer peer review. If you know someone like that, please either reply on this thread or find me on Mastodon: https://hachyderm.io/@manchicken

### What happened to the Rust rewrite idea?

Given my ability to manage my own schedule, I don't think that there's any way that I can produce a decent version any time soon, and it has already been too long since I updated the module. The last week has me convinced that the C bindings are still good, the code just needs a bunch of cleanup.

### New Release Timeline

I'm hoping to have a new version out by the end of this month, perhaps even as soon as this weekend. It really depends on whether I find any more surprises in the toolchain as I did last time I tried to prep a maintenance release.

### Conclusion

I'm going to close this thread, as I'm feeling a lot more optimistic now that I've got some new ideas on the matter. Thanks, and please let me know if you have any more thoughts.
